src/dotnet-uninstall/Shared/BundleInfo/BundleVersion.cs

```diff
@@ -0,0 +1,109 @@
+﻿using System;
+using Microsoft.DotNet.Tools.Uninstall.Shared.Exceptions;
+using Microsoft.DotNet.Tools.Uninstall.Shared.Utils;
+
+namespace Microsoft.DotNet.Tools.Uninstall.Shared.BundleInfo
+{
+    internal class BundleVersion : IEquatable<BundleVersion>, IComparable<BundleVersion>, IComparable
+    {
+        public int Major { get; }
+        public int Minor { get; }
+        public int Patch { get; }
+        public int? Preview { get; }
+
+        private readonly string _versionString;
+
+        public BundleVersion(string versionString)
+        {
+            var regex = Regexes.DotNetCoreBundleVersionRegex;
+            var match = regex.Match(versionString);
+
+            if (!match.Success)
+            {
+                throw new InvalidVersionStringException(versionString);
+            }
+
+            var versionMajorString = match.Groups[Regexes.VersionMajorGroupName].Value;
+            var versionMinorString = match.Groups[Regexes.VersionMinorGroupName].Value;
+            var versionPatchString = match.Groups[Regexes.VersionRegexPatchGroupName].Value;
+
+            var versionPreviewString = match.Groups["preview"].Success ? match.Groups["preview"].Value : null;
+
+            Major = int.Parse(versionMajorString);
+            Minor = int.Parse(versionMinorString);
+            Patch = int.Parse(versionPatchString);
+            Preview = versionPreviewString != null ? int.Parse(versionPreviewString) as int? : null;
+
+            _versionString = versionString;
+        }
+
+        public override string ToString()
+        {
+            return _versionString;
+        }
+
+        public override bool Equals(object obj)
+        {
+            return Equals(obj as BundleVersion);
+        }
+
+        public bool Equals(BundleVersion other)
+        {
+            return other != null &&
+                   Major == other.Major &&
```

Comment:
Just checking _versionString is not enough, after the null check?

Reply:
I don't think so, because different strings may correspond to the same version (there are different version string formats for SDK and Runtime). So even if two `_versionString`s are different, the versions they correspond to are the same, e.g. `3.0.100 - preview5` and `3.0.1 Preview 5`.